### PR TITLE
[SDPA] Update sdpa dispatch logic to enable fused backward

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Reorganizes how the sdp dispatch logic is down in order to enable backwards for fused kernels

### CODE REVIEW

File: test/test_transformers.py [hunk 1]
```diff
@@ -1118,7 +1123,39 @@ def test_efficient_attention_gradcheck(self, contiguous_inputs: bool):
         # in fp32
         assert gradcheck(lambda *args, **kwargs:
                          wrapper_set_seed(torch.ops.aten._efficient_attention_forward, *args, **kwargs),
-                         (query, key, value, None, None, None, True, False), fast_mode=True, atol=8e-5, rtol=1e-3)
+                         (query, key, value, None, None, None, True, False), fast_mode=False, atol=8e-5, rtol=1e-3)
```
Comment (drisspg) [inline test/test_transformers.py:1126]: @albanD This test if passing with fast_mode = True. The new test I wrote below uses this torch.ops.aten function in it's backward. Those tests also fail in slow mode but pass with fast_mode = True. The kerenls don't have a fp64 implementation that is why it is done in torch.float32. 

All this context to say, from reading the fast_mode=True in the docs it doesn't say one mode is more accurate than the other. However I would assume slow is more accurate. How would I proceed in verifying this computed grad is correct?

@cpuhrsch @danthe3rd

File: test/test_transformers.py [hunk 2]
```diff
@@ -1144,7 +1181,7 @@ def test_fused_sdp_choice(self, type: str):
             value = value.view(batch_size, -1, num_heads, head_dim).transpose(1, 2)
             key = key.view(batch_size, -1, num_heads, head_dim).transpose(1, 2)
 
-            if SM80OrLater:
+            if SM80OrLater and not type == "nested":
```
Comment (drisspg) [inline test/test_transformers.py:1184]: flash-attention for nested tensors is currently turned of because of cuda memory issues

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 3]
```diff
@@ -232,14 +259,15 @@ inline bool use_mem_efficient_attention(sdp_params params, bool debug) {
       at::kHalf, at::kFloat, at::kBFloat16};
 
   //  Define gate functions that determine if a flash kernel can be ran
-  std::vector<std::function<bool(sdp_params, bool)>> constraints{
+  constexpr std::array<bool(*)(sdp_params, bool), 8> constraints{{
       check_gpu_sm50_or_greater,
       check_runtime_disabled_mem_efficient,
+      check_requires_grad_and_nested,
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:265]: Since the caller is composite we add a check to ensure that if the input is nested and requires grad we don't return fused kernel.

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 4]
```diff
@@ -198,7 +201,7 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> _efficient_attention_backward(
     ASSIGN_CHECK_OVERFLOW(p.gQ_strideH, grad_q.stride(2));
     ASSIGN_CHECK_OVERFLOW(p.gK_strideH, grad_k.stride(2));
     ASSIGN_CHECK_OVERFLOW(p.gV_strideH, grad_v.stride(2));
-    p.gQKV_strideM_multiplier = grad_q.is_contiguous() ? 1 : 3;
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:201]: There are many ways that grad_q can be non-contiguous and empty_like will preserve the strides of input.

File: test/test_transformers.py (hunk 1 above)
Comment (drisspg) [reply test/test_transformers.py:1126]: I think noteworthy is that the "math" kernel also fails with fast_mode=False in fp32 and these tolerances

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 5]
```diff
@@ -492,16 +533,17 @@ Tensor flash_attention_helper(
   // If we are passing in query, key, value all the same tensors then we have
   // packed them into one tensor and need to slice for flash attention
   Tensor attention =
-      at::_flash_scaled_dot_product_attention(
+      std::get<0>(at::_flash_attention_forward(
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:536]: I updated this function but also not really sure if this even used anymore, in fact I would prefer if it wasn't since it feels vestigial.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 6]
```diff
@@ -432,22 +473,22 @@ std::tuple<Tensor, Tensor> mem_efficient_helper_nested_unpacked(
       {Nnz_kv, num_heads, head_dim},
       {nnz_v_stride, head_v_stride, head_dim_stride},
       value_impl->get_storage_offsets()[0]);
-  std::tuple<Tensor, Tensor> attention_and_weights =
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:435]: its not weights but the output, and logsumexp

File: aten/src/ATen/native/native_functions.yaml [hunk 7]
```diff
@@ -13244,18 +13244,39 @@
...
+    CUDA: _scaled_dot_product_efficient_attention_backward_cuda
+
+# Returns ouput, softmax_logsumexp, softmax
+- func: _flash_attention_forward(Tensor query, Tensor key, Tensor value, Tensor cum_seq_q, Tensor cum_seq_k, int max_q, int max_k, bool return_softmax, float dropout_p, bool is_causal) -> (Tensor, Tensor, Tensor)
   variants: function
+  dispatch:
+    CUDA: _flash_attention_forward
+
+# Returns ouput, logsumexp if compute_logsumexp
+- func: _efficient_attention_forward(Tensor query, Tensor key, Tensor value, Tensor? cu_seqlens_q, Tensor? cu_seqlens_k, int? max_seqlen_q, bool compute_log_sumexp=False, bool causal=False) -> (Tensor, Tensor)
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:13271]: This can really be subsumed into the both the forward calls of  _scaled_dot_product_efficient_attention for cuda and nested tensor. Same for the backwards. However for developing and bisecting where errors can/do get introduced it was helpful to have this separated by a dispatcher call. Seems like a nice follow up PR

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -12003,6 +12003,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
             # This is only failing on Linux Bionic 3.10 Cuda 11.6
             DecorateInfo(unittest.skip("Skipped!"), 'TestCommon', 'test_dtypes',
                          device_type='cuda', active_if=_get_torch_cuda_version() >= (11, 6)),
+            DecorateInfo(unittest.skip("Skipped!"), 'TestCommon', 'test_noncontiguous_samples',
```
Comment (drisspg) [inline torch/testing/_internal/common_methods_invocations.py:12006]: This is only failing because this calls efficient_attention which can handle all non-contiguity but the last dim not ==1. So I added the skip here but also could add another gate so that we go down the math part. But I feel like should fail loudly in this case and someone
can only enable the math kernel

File: test/test_transformers.py (hunk 1 above)
Comment (soulitzer) [reply test/test_transformers.py:1126]: A case where fast mode may fail to catch incorrect gradients is the case when a few elements of the Jacobian are wrong, but the rest is correct. (In contrast, if we had something that was element-wise, fast mode would very likely catch this failure).

If you're only able to run in fp32, comparing the size of the error to what you get when running forward as a composite operator is a good proxy though.

**Review by cpuhrsch — APPROVED** [review]
